keepassxreboot/keepassxc
# Add favicon fetch button next to entry's url edit textbox

## Description

When WITH_XC_NETWORKING is defined, create a QToolButton beside
the Edit Entry -> Entry -> URL which when pressed, acts as though the
Edit Entry -> Icon -> Download Favicon button is pressed.  This button
is disabled (grayed-out) when the URL text is empty, and enabled when
the text is present. This PR also removes the `UrlFetchProgessDialog` class
in favor of a "Custom icon successfully download" message (if triggered from
the edit entry tab).

## Motivation and context


Fixes #936

## How has this been tested?



- Compile with `-DWITH_XC_NETWORKING=OFF`
  - Start to create new entry "Favicon Fetch Entry", verify no button present next to url box 
  ![2018-10-30_22-58-00](https://user-images.githubusercontent.com/5200550/47771632-52b15980-dca1-11e8-8c42-3793d833736d.png)
  - Enter URL in text box, verify still no button
- Compile with `-DWITH_XC_ALL=ON`
  - Start to create new entry "Favicon Fetch Entry (Networking Enabled)", verify icon button is present, yet disabled 
![2018-11-18_19-25-36_no-url](https://user-images.githubusercontent.com/5200550/48684778-2bdf9800-eb68-11e8-9d69-80d6303f5701.png)
  - Enter text in the URL box and verify the button becomes clickable 
![2018-11-18_19-26-30_with-url](https://user-images.githubusercontent.com/5200550/48684783-326e0f80-eb68-11e8-8df7-76d49d40b644.png)
  - Click button.  ~Verify progress dialogs appear and no error is present.~ _(removed in this PR)_. Verify "Custom icon successfully downloaded" message appears.
![msg](https://user-images.githubusercontent.com/5200550/48684835-6f3a0680-eb68-11e8-8237-c2acde75bd8a.png)
  - Go to Icon tab and verify presence of correct favicon 
![2018-10-30_23-08-33](https://user-images.githubusercontent.com/5200550/47771853-15010080-dca2-11e8-9415-ed6831353f96.png)
  - Return to main entry edit tab and click button again, verify that "Custom icon already exist" info message appears 
![already](https://user-images.githubusercontent.com/5200550/47771898-3c57cd80-dca2-11e8-8c16-ec02291051b7.png)
  - Change URL to nonexistent URL and attempt to fetch icon for it.  Verify "Unable to fetch favicon" error message appears, and last correct favicon still set.
![2018-10-30_23-08-00](https://user-images.githubusercontent.com/5200550/47771955-67422180-dca2-11e8-9c7e-b450be3926e0.png)
  - Save entry and verify that it has new icon in database widget 
![2018-10-30_23-09-21](https://user-images.githubusercontent.com/5200550/47772033-9c4e7400-dca2-11e8-886a-082e35f52569.png)
- Run existing tests both with and without networking.


## Types of changes


- ✅ New feature (non-breaking change which adds functionality)

## Checklist:




- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
